daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1593,11 +1593,23 @@ private[lf] object SBuiltin {
           // TODO (drsk) validate key hash. https://github.com/digital-asset/daml/issues/13897
           machine.disclosureTable.contractIdByKey.get(gkey.hash) match {
             case Some(coid) =>
-              val vcoid = coid.value
-              val (control, _) = continue(Some(vcoid))
-              control
+              machine.disclosureTable.contractById.get(coid) match {
+                case Some((actualTemplateId, _)) if actualTemplateId == operation.templateId =>
+                  val vcoid = coid.value
+                  continue(Some(vcoid))._1
+
+                case Some((actualTemplateId, _)) =>
+                  throw SErrorDamlException(
+                    IE.WronglyTypedContract(coid.value, operation.templateId, actualTemplateId)
```

Comment:
Throwing an error here definitely makes sense. I do wonder though if this is the right error:

Consider the current cases where we throw WronglyTypedContract: They’re pretty much all user errors. E.g., I called a choice on a contract of a different type. The crucial part is the reason why we get this error is that the participant knows the correct contractid -> contract mapping and that does not match how you tried to use it.

Now here the situation is a bit different. It's not my usage that is wrong, the mapping in the disclosure table is wrong: A key hash for template T maps to a contract of template T'.

So what do you think of throwing something like an `InconsistentDisclosureTable` error or something along those lines to give users more information on what might have gone wrong and how they can fix it?